### PR TITLE
feat: implement getting address of function

### LINKED ISSUE #181: Getting address of function is not supported
```c
int foo(int t)
{
    return t + 2;
}

typedef int (*foo_t)(int x);

int main(void)
{
    foo_t x = &foo;
    return x(40);
}
```
produce
```
Unhandled exception. System.NotSupportedException: Cannot find variable foo.
   at Cesium.CodeGen.Ir.Expressions.IdentifierExpression.Resolve(IDeclarationScope scope) in C:\d\github\Cesium\Cesium.CodeGen\Ir\Expressions\IdentifierExpression.cs:line 42
   at Cesium.CodeGen.Ir.Expressions.UnaryOperatorExpression.<EmitTo>g__EmitGetAddress|5_1(IDeclarationScope scope, IExpression target) in C:\d\github\Cesium\Cesium.CodeGen\Ir\Expressions\UnaryOperatorExpression.cs:line 54
   at Cesium.CodeGen.Ir.Expressions.UnaryOperatorExpression.EmitTo(IDeclarationScope scope) in C:\d\github\Cesium\Cesium.CodeGen\Ir\Expressions\UnaryOperatorExpression.cs:line 34
   at Cesium.CodeGen.Ir.BlockItems.DeclarationBlockItem.EmitScopedIdentifier(IDeclarationScope scope, ScopedIdentifierDeclaration scopedDeclaration) in C:\d\github\Cesium\Cesium.CodeGen\Ir\BlockItems\DeclarationBlockItem.cs:line 90
   at Cesium.CodeGen.Ir.BlockItems.DeclarationBlockItem.EmitTo(IDeclarationScope scope) in C:\d\github\Cesium\Cesium.CodeGen\Ir\BlockItems\DeclarationBlockItem.cs:line 50
   at Cesium.CodeGen.Ir.BlockItems.CompoundStatement.EmitTo(IDeclarationScope scope) in C:\d\github\Cesium\Cesium.CodeGen\Ir\BlockItems\CompoundStatement.cs:line 28
   at Cesium.CodeGen.Ir.TopLevel.FunctionDefinition.EmitCode(TranslationUnitContext context, FunctionScope scope) in C:\d\github\Cesium\Cesium.CodeGen\Ir\TopLevel\FunctionDefinition.cs:line 229
   at Cesium.CodeGen.Ir.TopLevel.FunctionDefinition.EmitTo(TranslationUnitContext context) in C:\d\github\Cesium\Cesium.CodeGen\Ir\TopLevel\FunctionDefinition.cs:line 87
   at Cesium.CodeGen.Contexts.AssemblyContext.EmitTranslationUnit(IEnumerable`1 nodes) in C:\d\github\Cesium\Cesium.CodeGen\Contexts\AssemblyContext.cs:line 49
   at Cesium.Compiler.Compilation.GenerateCode(AssemblyContext context, String inputFilePath) in C:\d\github\Cesium\Cesium.Compiler\Compilation.cs:line 107
   at Cesium.Compiler.Compilation.GenerateCode(AssemblyContext context, String inputFilePath) in C:\d\github\Cesium\Cesium.Compiler\Compilation.cs:line 107
   at Cesium.Compiler.Compilation.Compile(IEnumerable`1 inputFilePaths, String outputFilePath, TargetRuntimeDescriptor targetRuntime, Nullable`1 moduleKind, String namespace, String globalClassFQN) in C:\d\github\Cesium\Cesium.Compiler\Compilation.cs:line 29
```

### PR DESCRIPTION
**Pull request by FrediKats**
Try to close #181 

I am not sure about init getting LValueLocalVariable for func. FunctionInfo and FunctionType is not LValue. And... shouldn't be, is it?  But I don't know other way of implementation.

p.s. thx @kant2002 for FunctionType.

### CODE REVIEW

File: Cesium.CodeGen/Ir/Expressions/IdentifierExpression.cs [hunk 1]
```diff
@@ -32,27 +35,44 @@ public IdentifierExpression(Ast.IdentifierExpression expression)
...
-                throw new NotSupportedException(
-                    $"Variable {Identifier} is both available as a local and as a function parameter.");
+            return new LValueParameter(par);
         }
+        
+        if (fun is not null)
+        {
+            var functionType = new FunctionType(fun.Parameters, fun.ReturnType);
+            var variableDefinition = new VariableDefinition(functionType.ResolvePointer(scope.Context));
+            return new LValueLocalVariable(variableDefinition);
```
Comment (ForNeVeR) [inline Cesium.CodeGen/Ir/Expressions/IdentifierExpression.cs:66]: Free function shouldn't be represented by `LValueLocalVariable`; creating a `VariableDefinition` in this point is wrong.

Moreover, after I've taken a look at the code using `IdentifierExpression`, I now think that maybe a better abstraction should be used.

Let's create a new interface, `IValue` (without the `EmitSetValue` method), and inherit `ILValue` from it.

`IdentifierExpression::Resolve` would then return an `IValue` and not an `ILValue`. And for a function, let's create a new implementation of `IValue`.

File: Cesium.CodeGen/Ir/Expressions/IdentifierExpression.cs [hunk 2]
```diff
@@ -32,27 +35,44 @@ public IdentifierExpression(Ast.IdentifierExpression expression)
...
+
+        }
+
+        if (globalType != null)
+        {
+            var flobalField = scope.Context.AssemblyContext.ResolveGlobalField(Identifier, scope.Context);
```
Comment (ForNeVeR) [inline Cesium.CodeGen/Ir/Expressions/IdentifierExpression.cs:72]:
```suggestion
            var globalField = scope.Context.AssemblyContext.ResolveGlobalField(Identifier, scope.Context);
```

File: Cesium.CodeGen/Ir/Expressions/IdentifierExpression.cs [hunk 3]
```diff
@@ -32,27 +35,44 @@ public IdentifierExpression(Ast.IdentifierExpression expression)
...
+            return new LValueGlobalVariable(flobalField);
+        }
+
+        throw new NotSupportedException($"Cannot find {Identifier} into variable, parameter of function.");
```
Comment (ForNeVeR) [inline Cesium.CodeGen/Ir/Expressions/IdentifierExpression.cs:76]:
```suggestion
        throw new NotSupportedException($"Cannot find a local variable, a function parameter, a global variable or a function {Identifier}.");
```

File: Cesium.IntegrationTests/pointer_to_func.c [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    return t + 2;
+}
+
+typedef int (*foo_t)(int x);
+
+int main(void)
+{
+    foo_t x = &foo;
+    return 40;
+}
```
Comment (ForNeVeR) [inline Cesium.IntegrationTests/pointer_to_func.c:12]: Let's also create a unit test to verify the resulting IL code?

**Review by ForNeVeR — CHANGES_REQUESTED** [review]
That's a nice start, but I think that we should improve our architecture around lvalues and rvalues. See my explanation in the comments to the code.

File: Cesium.CodeGen/Ir/Expressions/IdentifierExpression.cs (hunk 1 above)
Comment (FrediKats) [reply Cesium.CodeGen/Ir/Expressions/IdentifierExpression.cs:66]: Thx. I was thinking about this. That is what I mean in my `shouldn't be, is it?`. Will try to do it.

File: Cesium.CodeGen/Ir/Expressions/AssignmentExpression.cs [hunk 5]
```diff
@@ -45,7 +46,7 @@ public override void EmitTo(IDeclarationScope scope)
         if (Operator != BinaryOperator.Assign)
             throw new NotSupportedException($"Operator {Operator} should've been lowered before emitting.");
 
-        _target.Resolve(scope).EmitSetValue(scope, Right);
+        ((ILValue)_target.Resolve(scope)).EmitSetValue(scope, Right);
```
Comment (FrediKats) [inline Cesium.CodeGen/Ir/Expressions/AssignmentExpression.cs:49]: I am not sure about this place.

File: Cesium.CodeGen/Ir/Expressions/FunctionPointerValue.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    private readonly MethodReference _methodReference;
+
+    public FunctionPointerValue(MethodReference methodReference)
+    {
+        _methodReference = methodReference;
+    }
+
+    public void EmitGetValue(IDeclarationScope scope)
+    {
+        throw new NotImplementedException();
```
Comment (FrediKats) [inline Cesium.CodeGen/Ir/Expressions/FunctionPointerValue.cs:18]: I need some help with implementation of this method.

File: Cesium.CodeGen/Ir/Expressions/SubscriptingExpression.cs [hunk 7]
```diff
@@ -30,11 +30,11 @@ public IExpression Lower()
...
     public TypeReference GetExpressionType(IDeclarationScope scope) => Resolve(scope).GetValueType();
 
-    public ILValue Resolve(IDeclarationScope scope)
+    public IValue Resolve(IDeclarationScope scope)
     {
         if (_expression is not IdentifierExpression identifier)
             throw new NotImplementedException("Subscription supported only for IdentifierConstantExpression");
 
-        return new LValueArrayElement(identifier.Resolve(scope), _index);
+        return new LValueArrayElement((ILValue)identifier.Resolve(scope), _index);
```
Comment (FrediKats) [inline Cesium.CodeGen/Ir/Expressions/SubscriptingExpression.cs:38]: I am not sure about this place.

File: Cesium.CodeGen/Ir/Expressions/FunctionPointerValue.cs (hunk 6 above)
Comment (ForNeVeR) [reply Cesium.CodeGen/Ir/Expressions/FunctionPointerValue.cs:18]: Alright; I'm on it.

**Review by ForNeVeR — APPROVED** [review]
Thanks a lot!

**Pull request merged**